[이지예] 숫자 야구 게임 구현 완료!

jiyeeeah opened the pull request:
## ⚾ Baseball 기능 구현 목록

### 📌 게임 시작

- [x] 게임 시작 문구 **출력**
- [x] 컴퓨터의 숫자 생성 기능
  - [x] 1 ~ 9까지 중 겹치지 않는 3개의 숫자

### 📌 게임 진행

- [x] 사용자 **입력**
  - [x] 입력 받은 값의 validation 확인
  - [x] 문자가 아닌 '숫자'인지
  - [x] 3자리의 수인지
  - [x] [ERROR] 입력값이 유효하지 않은 경우, throw 문으로 종료
- [x] 볼, 스트라이크, 낫싱 결과 문구 **출력**
- [x] 맞추면 게임 종료 문구 출력

### 📌 게임 종료

- [x] 게임 종료 후, 재시작 여부 물어보는 입력 받기
  - [x] 1이나 2를 받았는지 validation 확인
- [x] 입력값이 1일 경우, 게임 재시작

## ⚾ Baseball 코드 구성
### `App.js`
- 메인 코드(`BaseballGame.js`)가 실행되는 곳입니다.
### `BaseballGame.js`
  - `BaseballGame` class를 담고 있는 곳입니다.
- `ComputerNums`, `UserNums`
  - 컴퓨터가 선택한 숫자의 배열과 사용자가 입력한 숫자의 배열을 갖습니다.
  - 둘 다 setter를 갖습니다.
    - 특히, `UserNums`의 setter에는 유효성 체크를 위한 util 함수가 포함되어 있습니다.
- `COMPUTER_NUM_LEN`
  - 야구 게임의 길이를 3으로 고정시키기 위해 COMPUTER_NUM_LEN을 선언했습니다.
- `runGame()`
  - 전체 게임의 프로세스를 담았습니다. (재시작할 때 용이하기 위해)
- `startGame()`
  - 게임의 시작 부분입니다.
  - [x] 게임 시작 멘트 출력
  - [x] 컴퓨터가 숫자 선택
- `playGame()`
  - 게임이 진행되는 부분입니다.
  - `isUserWin`을 flag로 하여, 3스트라이크가 나올 때까지 while문으로 입력을 반복합니다.
  - 입력이 반복되면서 `UserNums`가 바뀌고, 볼/스트라이크/낫싱의 결과문을 출력합니다.
- `endGame()`
  - 게임 종료시 재시작을 묻는 질문과 함께 플레이어의 입력을 받습니다.
  - 재시작 선택 시, `runGame()`을 다시 실행합니다.

jiyeeeah committed:
[docs] 기능 구현 목록 작성

jiyeeeah committed:
[feat] baseball 게임 기능 구현

jiyeeeah committed:
[docs] 기능 구현 목록 체크 완료

jiyeeeah committed:
[docs] util 함수들의 주석 추가

wo-o29 commented on docs/README.md:
요구 사항이 디테일하네요👍👍👍 많이 배워갑니다...!
근데 혹시 에러 메세지에 `[ERROR]`를 추가해야 한다는 요구사항은 테스트 코드를 분석 하신건가요??

jiyeeeah replied:
네 맞아요.. jest는 처음 보는거라서 좀 헤매고 검색의 도움도 좀 받았습니다 ㅎㅎ

wo-o29 replied:
저도 테스트 코드 경험이 없는데 다음주차부터는 테스트 코드를 짜는 것 같아서 빨리 배워야겠네요...

jiyeeeah replied:
저도 없어서.. 남는 시간동안 공부 좀 해보려구요! 이번 문제에서 test 코드 짜보기도 도전해보려고 합니다 :fist_right:

wo-o29 commented on src/BaseballGame.js:
```diff
@@ -0,0 +1,74 @@
+   * 게임 시작 멘트를 출력하며, 컴퓨터의 숫자 3개를 정합니다.
+   */
+  async startGame() {
+    await MissionUtils.Console.print("숫자 야구 게임을 시작합니다.");
```
```suggestion
    await MissionUtils.Console.print("숫자 야구 게임을 시작합니다.");
```
`MissionUtils.Console.print` 메서드는 Promise를 반환하지 않는 동기 함수여서 `await`에 영향을 받지 않아서 제거해줘도 좋을 것 같아요!

jiyeeeah replied:
헐랭... `MissionUtils.Console.readLineAsync`땜에 한번에 다 `await`를 넣어준 것 같네요ㅠㅠ 좀 더 꼼꼼하게 확인해볼걸 그랬습니다. 감사드립니다!
혹시 `MissionUtils.Console.print`가 동기함수인건 어떻게 아셨는지 여쭤봐도 될까요?

wo-o29 replied:
`node_modules/@woowacourse/mission-utils/src/console.js` 라이브러리 패키지 내부 코드에 `print` 메서드는 Promise를 반환하지 않고 단순한 `console.log` 메서드인걸로 확인했습니다!

```js
class Console {
  constructor() {}

  static print(message) {
    console.log(message);
  }
}
```

jiyeeeah replied:
:open_mouth: 와우 감사합니다!

wo-o29 commented on src/BaseballGame.js:
```diff
@@ -0,0 +1,74 @@
+
+      this.setUserNums(USER_INPUT);
+
+      const { strike, ball } = await getResult({
+        computerNums: this.#ComputerNums,
+        userNums: this.#UserNums,
+      });
```
```suggestion
      const { strike, ball } = await getResult(this.#ComputerNums, this.#UserNums);
```
인자 넘기실 때 객체로 넘겨주시는 이유가 있으신가요??  
그리고 `getResult` 함수는 비동기적으로 동작하지 않아서 `await`을 사용하지 않아도 될 것 같습니다!

jiyeeeah replied:
혹시 `computerNums`와 `UserNums`의 순서가 바뀌거나 할 수 있어서 이런식으로 작성했습니다.
근데 지금 다시 생각해보니 다른 함수들과는 통일성이 없어서 아쉽네요.. 다른 함수들도 객체로 넘기든가 얘네도 그냥 파라미터로 넘기던가 하는게 좋을 수도 있을 것 같습니다. 이 부분은 리팩토링하면서 좀 더 생각해보겠습니다!
(+) 왤케 `await`을 남용한거지ㅠㅠ 더 꼼꼼하게 확인해야겠습니다.. 감사합니다!

wo-o29 replied:
클래스 문법이여서 그런지 지금 하신 방법도 가독성은 좋은 것 같아요!

wo-o29 commented on src/BaseballGame.js:
```diff
@@ -0,0 +1,74 @@
+   * 값을 입력 받고, 받은 값의 유효성 체크와 결과를 출력합니다.
+   */
+  async playGame() {
+    let isUserWin = false;
```
```suggestion
    let isUserWin = false;
```
while 문 중지 조건을 `break`를 사용하지 않으시고 `flag` 방식으로 하신 이유가 있으신가요??

jiyeeeah replied:
가독성을 위해 그렇게 했습니다! 
`break;`로 `while`문을 중지하려면 `while`의 조건 부분에 `(true)`라고 적어야하는데 개인적으로 선호하지 않아서요..

wo-o29 replied:
오호 그렇군요 좋은 것 같습니다!

wo-o29 commented on src/utils/getResult.js:
```diff
@@ -0,0 +1,35 @@
+ * @param {number[]} params.userNums - 플레이어가 입력한 숫자 배열
+ * @returns {{ strike: number, ball: number }} - 스트라이크와 볼 개수
+ */
+export async function getResult({ computerNums, userNums }) {
```
```suggestion
export async function getResult({ computerNums, userNums }) {
```
`getResult` 함수 내부에 비동기 함수가 없는 것 같은데 `async` 함수로 선언하신 이유가 있으신가요??

jiyeeeah replied:
ㅋ.... 그러게요ㅠㅠ 코드를 이리저리 수정하다가 놓친것 같습니다ㅠ 감사합니다!

wo-o29 commented on src/BaseballGame.js:
```diff
@@ -0,0 +1,74 @@
+      "게임을 새로 시작하려면 1, 종료하려면 2를 입력하세요. \n"
+    );
+    isEndInputValid(USER_INPUT);
+    if (USER_INPUT === "1") await this.runGame();
+    else await MissionUtils.Console.print("게임이 종료되었습니다.");
```
```suggestion
    if (USER_INPUT === "1"){ 
       await this.runGame();
       return;
    }
    
    MissionUtils.Console.print("게임이 종료되었습니다.");
```
저는 이런 경우에 `early return`을 사용하려고 하는 것 같아요! 이건 취향 차이일 것 같네요🙃
그리고 `MissionUtils.Console.print` 메서드도 `await`에 영향이 가지 않을 것 같아요

jiyeeeah replied:
오 여기에 `return`을 넣으시는 이유가 뭔가용? :thinking:

wo-o29 replied:
`else`를 생략할 수 있어서 선호하는 것 같습니다. early return을 사용하면 `depth`가 늘어나지 않아서 좋은 것 같아요!

jiyeeeah replied:
혹시 `return`없이도 `else`는 생략 가능하지 않나요? 지금 이 경우에서는 depth가 늘어나지도 않구요! 
``` suggestion
    if (USER_INPUT === "1") await this.runGame();
    await MissionUtils.Console.print("게임이 종료되었습니다.");
```
오히려 이렇게 표현해도 되지 않을까요?
`return`을 적어줘야 함수 탈출이 명확하려나요? :thinking:

wo-o29 replied:
> 혹시 `return`없이도 `else`는 생략 가능하지 않나요? 지금 이 경우에서는 depth가 늘어나지도 않구요!

-> 맞습니다! 지금 코드에서 early return의 이점은 `else`문 생략일 것 같은데, 추후에 요구사항이 늘어나서 입력 값이 `2`이고 `3번 안에 3스트라이크를 맞춘 경우`와 같이 조건이 늘어나면 depth가 늘어날 요소가 생길 것 같아서 early return을 사용하면 유지보수하기 편할 것 같아요!

> 오히려 이렇게 표현해도 되지 않을까요? `return`을 적어줘야 함수 탈출이 명확하려나요? 🤔

-> return을 적어주지 않으면 첫 번째 게임이 끝나지 않은 상태로 두 번째 게임이 시작되어서 게임 종료될 때 종료 문구가 중첩되어서 나올 것 같아요! 그래서 return으로 첫 번째 게임을 끝내주고 두 번째 게임으로 들어가야 할 것 같아요

이렇게 적으면 조금 이해하기 어려우실 것 같아서 추가 설명 드릴게요! 혹시라도 이해가 잘 안가시면 더 물어보셔도 됩니다!!
```
초기 게임 시작(첫 번째 게임 시작)
3스트라이크 
유저 입력: 1 
게임 재시작(두 번째 게임 시작, 아직 첫 번째 게임은 끝나지 않은 상태 await 걸려있기 때문에  뒤에 게임이 끝나야 끝남)
3스트라이크
유저 입력: 2 (두 번째 게임 종료)
게임이 종료되었습니다. (두 번째 게임 종료 문구 출력)
게임이 종료되었습니다. (첫 번째 게임 종료 문구 출력)
```

wo-o29 replied:
추가적으로 지금 코드를 보다가 알았는데 예지님 코드에서 유효성 검사하는 부분의 인터페이스를 봤을 때 유효성 검사에 통과하지 못하면 어떤 일이 발생하는지 파악하려면 내부 코드에 들어가야 알 수 있는 것 같아요!
```js
  async endGame() {
    const USER_INPUT = await MissionUtils.Console.readLineAsync(
      "게임을 새로 시작하려면 1, 종료하려면 2를 입력하세요. \n"
    );
    
   // 1. 함수 이름은 유효성 검사를 하는 것 같은데 통과하지 못하면 어떻게 되는지 여기서는 파악하기 어렵다
    isEndInputValid(USER_INPUT);  // 기존 코드
    
   
   // 2. is, has, should 는 주로 boolean 값을 다룰 때 사용하는 네이밍인 것 같아서 내부에서 에러를 throw하지 않고 boolean 반환
   // 3. 반환받은 boolean 값으로 에러를 throw 하면 endGame함수 인터페이스만 보고도 동작을 알 수 있다!
   const isEndInputValidate = isEndInputValid(USER_INPUT); 
   
   if(isEndInputValidate){
     throw new Error("에러 발생 ㅇ_ㅇ!")
   }

   // another code...
  }
```
이렇게 유효성 검사 함수에서 `boolean` 값을 반환받고 그 값에 따라 에러를 `throw`하는 로직은 어떻게 생각하시나요?!

jiyeeeah replied:
아 그렇네요..! 당연한건데 생각지도 못했어요..! `return`을 써주는게 맞겠네요! 감사합니다 :+1: 
유효성 검사 관련된 내용도 리팩토링 하면서 더 고민해보겠습니다! 감사해요 :relaxed:

wo-o29 commented on src/BaseballGame.js:
```diff
@@ -0,0 +1,74 @@
+      });
+      await printResult({ strike, ball });
+      if (strike === 3) {
+        await MissionUtils.Console.print("3개의 숫자를 모두 맞히셨습니다! 게임 종료");
```
```suggestion
        MissionUtils.Console.print("3개의 숫자를 모두 맞히셨습니다! 게임 종료");
```
요기도 `await` 제거해두 될 것 같습니다~!

wo-o29 commented on src/BaseballGame.js:
```diff
@@ -0,0 +1,74 @@
+        computerNums: this.#ComputerNums,
+        userNums: this.#UserNums,
+      });
+      await printResult({ strike, ball });
```
```suggestion
      printResult({ strike, ball });
```
요기도 await 제거해두 될 것 같습니다~!

wo-o29 commented on src/utils/generateRandomNum.js:
```diff
@@ -0,0 +1,20 @@
+ * @param {number} length - 컴퓨터가 선택하는 수의 개수
+ * @returns {number[]} 컴퓨터가 선택한 수의 배열
+ */
+export async function generateRandomNum(min, max, length) {
+  const COMPUTER_NUMS = [];
+  while (COMPUTER_NUMS.length < length) {
+    const number = await MissionUtils.Random.pickNumberInRange(min, max);
+    // 중복되는 숫자 없도록 예외 처리
+    if (!COMPUTER_NUMS.includes(number)) {
+      COMPUTER_NUMS.push(number);
+    }
+  }
+  return COMPUTER_NUMS;
+}
```
`generateRandomNum` 함수도 비동기적으로 동작하는 요소가 없는 것 같아서 내부에 `async/await`제거해도 될 것 같습니다!

wo-o29 commented on src/utils/getResult.js:
```diff
@@ -0,0 +1,35 @@
+ * @param {number} params.strike - 스트라이크 개수
+ * @param {number} params.ball - 볼 개수
+ */
+export async function printResult({ strike, ball }) {
+  if (strike === 0 && ball === 0) await MissionUtils.Console.print("낫싱");
+  else if (strike === 0) await MissionUtils.Console.print(`${ball}볼`);
+  else if (ball === 0) await MissionUtils.Console.print(`${strike}스트라이크`);
+  else await MissionUtils.Console.print(`${ball}볼 ${strike}스트라이크`);
+}
```
`printResult` 함수도 `async/await` 제거해도 될 것 같습니다 ㅇ_ㅇ
결과 출력 방식이 저랑 약간 다르군요!
저는 `ball`, `strike`만 있는 경우를 따로 처리하지 않았던 것 같아요
`낫싱`, `3스트라이크`, `X볼 X스트라이크` 이렇게 세 개의 조건으로 했던 것 같네요
지예님 방식도 직관적이여서 좋은 것 같네요👍👍👍

jiyeeeah replied:
그렇게 하면 `0볼 1스트라이크` 이런식의 답변도 발생되지 않나용? :open_mouth:

wo-o29 replied:
> 그렇게 하면 `0볼 1스트라이크` 이런식의 답변도 발생되지 않나용? 😮

요기 부분은 제가 잘 이해하지 못한 것 같습니다... 제가 한 방식을 말씀드리면 `ball`, `strike` 카운트는 Default가 0이라서 이렇게 나옵니다!

```js
export const resultPrint = (strike, ball, nothing) => {
  // nothing early return
  if (nothing) {
    MissionUtils.Console.print(NOTHING_TEXT);
    return;
  }

  // 3strike early return
  if (strike === 3) {
    MissionUtils.Console.print(`3${STRIKE_TEXT}`);
    MissionUtils.Console.print(THREE_STRIKE_MESSAGE);
    MissionUtils.Console.print(RESTART_MESSAGE);
    return;
  }

  MissionUtils.Console.print(`${ball}${BALL_TEXT} ${strike}${STRIKE_TEXT} `);
};
```
<img width="468" alt="스크린샷 2024-09-20 오전 10 14 44" src="https://github.com/user-attachments/assets/2aa97c15-d6f3-4ef0-ac79-8b43a5ba413e">

jiyeeeah replied:
아하 어차피 테스트 돌릴 때 `1스트라이크` 라는 문구가 포함되어있으면 되니까 상관없겠군요!

저는 지금 올려주신 캡쳐의 `0볼 1스트라이크` 대신 `1스트라이크`만 나오도록 구현하고 싶어서 나눴습니다! (사실 꼭 그래야하는 줄 알았어요 ㅎㅎ)

wo-o29 replied:
오 그 생각은 못했는데 요구사항 보니깐 1스트라이크만 나오는게 있는 것 같네요 감사합니다!

wo-o29 commented on src/utils/validation.js:
```diff
@@ -0,0 +1,27 @@
+  if (INPUT_NUMS.length !== length) {
+    throw new Error(`[ERROR] 입력한 값이 ${length} 자리가 아닙니다.`);
+  }
+  for (let num of INPUT_NUMS) {
+    if (isNaN(num)) {
+      throw new Error("[ERROR] 숫자만 입력할 수 있습니다.");
+    }
+  }
```
```suggestion
    if (isNaN(numString)) {
      throw new Error("[ERROR] 숫자만 입력할 수 있습니다.");
    }
```
여기를 배열로 바꿔서 순회하시면서 개별로 처리해주셨는데 바로 입력받은 문자열에 `isNaN()` 하시는건 어떻게 생각하시나요??

jiyeeeah replied:
`numString`은 문자열이기 때문에 `isNaN(numString)`은 항상 `true`가 되지 않을까요?
입력받은 문자열 안의 각 문자들이 모두 숫자여야해서 반복문으로 일일이 확인해주었습니다!

wo-o29 replied:
제가 알기론 `isNaN()`은 인자로 받은 값을 Number 타입으로 변환해서 `NaN`인지 체크하는거라서 문제 없을 것 같아요!
대신 내부적으로 공백이랑 이런게 체크가 안되기는 하는데 저희는 입력 값 받을 때 유효성 검사를 거쳐서 이 문제는 괜찮을 것 같네요🙃
그래도 유효성 검사가 완벽히 통과되었다고 보장하기는 어려우니깐 지예님 방식도 좋은 것 같습니다!
![스크린샷 2024-09-20 오전 10 20 04](https://github.com/user-attachments/assets/eaa32fdc-7d32-4492-a766-f3456b138951)

jiyeeeah replied:
와.. 몰랐어요ㅠ 하나 또 배워갑니다! 정말 감사해요 :+1: :+1: :+1: 이런 단순한 것도 모르고 있었네요 저는..ㅎㅎ

wo-o29 commented on src/utils/getResult.js:
```diff
@@ -0,0 +1,35 @@
+  let ball = 0;
+  const ARR_LENGTH = Math.min(computerNums.length, userNums.length);
+
+  for (let i = 0; i < ARR_LENGTH; i++) {
+    for (let j = 0; j < ARR_LENGTH; j++) {
+      if (i === j && computerNums[i] === userNums[j]) strike++;
+      if (i !== j && computerNums[i] === userNums[j]) ball++;
+    }
+  }
```
```suggestion
  for (let i = 0; i < ARR_LENGTH; i++) {
    for (let j = 0; j < ARR_LENGTH; j++) {
      if (i === j && computerNums[i] === userNums[j]) strike++;
      if (i !== j && computerNums[i] === userNums[j]) ball++;
    }
  }
```
혹시 이렇게 이중 반복문으로 생각하신 이유가 있으신가요??
저는 루프는 한번 돌고 같은 자리이면 `strike`, `userNums[i]`가 `computerNums` 배열에 포함되어 있으면 `ball` 카운트해주었습니다

jiyeeeah replied:
`includes()` 메서드를 사용하신거죠? 그렇게도 생각해봤는데 `includes`도 어쨌든 배열을 순회하는 메서드이기 때문에 이중 반복문과 큰 차이는 없을거라고 생각했습니다!
근데 `includes()`를 사용하면 가독성이 좀더 좋아질 것 같기도 하네요! 참고하겠습니다. 감사합니당 :)

wo-o29 reviewed (approved):
전체적으로 코드가 잘 분리되어있고 JSDoc도 사용해 주셔서 흐름을 파악하기 좋았던 것 같네요👍👍

자바스크립트에서 클래스 문법까지 사용하시다니 굉장히 고수이시군요...!

혹시 코드 실행하실 때 에러가 발생하지는 않으셨나요? `ESModule`이기 때문에 `import` 할 때 확장자를 적어주지 않으면 에러가 발생할 것 같아서 여쭤봅니다!

그리고 이건 상수에 대한 관점에 따라서 조금 차이가 있을 것 같은데 지예님은 
```js
const COMPUTER_NUMS = [];
const INPUT_NUMS = numString.split("");
```
이러한 부분에도 스네이크 네이밍을 사용해 주셨는데 저는 저런 부분은 카멜케이스를 사용하고, 
```js
const DAYS = ["일", "월", "화", "수", "목", "금", "토"];
const END_MESSAGE = "게임을 종료합니다.";
```
이렇게 프로그램 실행 중에도 변하지 않는 부분에 대해서만 스네이크 네이밍을 사용하는 상수라고 생각했던 것 같아요

코드의 추상화 같은 부분이 너무 잘되어 있으셔서 많이 배웠던 것 같습니다!

저도 많이 놓치는 부분인데 커밋 단위를 조금 더 잘게 나눠보셔도 좋을 것 같아요!
그리고 전체적으로 `async/await` 사용 부분을 한번 확인해 보셔도 좋을 것 같습니다 수고하셨습니다🫡👏🙃☁️💡

wo-o29 commented on src/BaseballGame.js:
```diff
@@ -0,0 +1,74 @@
+
+  setUserNums(nums) {
+    isInputValid(nums, this.#COMPUTER_NUM_LEN);
+    this.#UserNums = nums.split("").map((str) => Number(str));
```
```suggestion
    this.#UserNums = nums.split("").map(Number);
```
사소하지만 요렇게 축약해도 동작은 같을 것 같아요 ㅇ_ㅇ!

jiyeeeah replied:
그러게요.. 찾아보니 그렇게도 가능하네요 하핳 왜 어려운 길을 가고 있었을까.. ㅋㅋㅋ
감사합니다!

wo-o29 replied:
그래도 지예님 코드가 더 목적성있고 의도가 보이는 것 같기도 하네요!

wo-o29 commented on src/utils/validation.js:
```diff
@@ -0,0 +1,27 @@
+ * return 값은 없으며, 유효하지 않을 경우 throw 문을 통해 예외를 발생시키고 프로그램을 종료합니다.
+ */
+export function isEndInputValid(numString) {
+  if (numString !== "1" && numString !== "2")
+    throw new Error("[ERROR] 입력하신 값이 유효하지 않습니다.");
```
```suggestion
  if (numString !== "1" && numString !== "2")
    throw new Error("[ERROR] 입력하신 값이 유효하지 않습니다.");
```
여기는 if문 내부가 1줄일 때 중괄호를 생략해 주셨는데, 위에 `isInputValid` 함수의 length validation 하실 때는 중괄호를 사용해 주셔서 하나로 통일해도 좋을 것 같아요🙃

jiyeeeah replied:
넹! 이런것까지 찾아주시다니 너무 감사드려요! ㅎㅎ

jiyeeeah commented:
@wo-o29 
> 혹시 코드 실행하실 때 에러가 발생하지는 않으셨나요? `ESModule`이기 때문에 `import` 할 때 확장자를 적어주지 않으면 에러가 발생할 것 같아서 여쭤봅니다!

=> 넴 오류 없더라구요! 확장자는 없는게 Airbnb 코드 컨벤션에서 추천하는 것 같구요!
참고 : https://github.com/woowacourse-precourse/javascript-baseball-6/pull/867/files#r1372454578


> 그리고 이건 상수에 대한 관점에 따라서 조금 차이가 있을 것 같은데 지예님은
> 
> ```js
> const COMPUTER_NUMS = [];
> const INPUT_NUMS = numString.split("");
> ```
> 
> 이러한 부분에도 스네이크 네이밍을 사용해 주셨는데 저는 저런 부분은 카멜케이스를 사용하고,
> 
> ```js
> const DAYS = ["일", "월", "화", "수", "목", "금", "토"];
> const END_MESSAGE = "게임을 종료합니다.";
> ```
> 
> 이렇게 프로그램 실행 중에도 변하지 않는 부분에 대해서만 스네이크 네이밍을 사용하는 상수라고 생각했던 것 같아요

=> 흠... 우혁님이 하신 방법이 조금 더 적합한 것 같습니다..! 참고해보도록 하겠습니다. 감사합니다!

리뷰 정말 자세하게 남겨주셔서 넘 감사드려요! `await`의 남용... 줄여보도록 하겠습니다 하핳 (민망함..)
그리구 커밋 단위도 일부러 한방에 했는데 나누는게 좋을 것 같군요..! (역시 커밋은 나눌수록 이득인건가) 
리뷰 정말 감사합니다!

wo-o29 commented:
> @jiyeeeah 
> => 넴 오류 없더라구요! 확장자는 없는게 Airbnb 코드 컨벤션에서 추천하는 것 같구요!
참고 : https://github.com/woowacourse-precourse/javascript-baseball-6/pull/867/files#r1372454578

오호 저런 리뷰가 있었군요...? 하지만 저희가 사용하는 환경은 바닐라 JS 환경이라서 일반적으로 개발할 때는 `webpack`이나 `vite`같은 번들러가 확장자를 생략해도 모듈을 찾을 수 있게 해주는 데 지금은 에러가 발생하는게 맞는 것 같아요! 

근데 올려주신 링크에서도 번들러는 사용하지 않는 환경인 것 같은데... 저 리뷰가 잘못되었을 가능성이 있을 것 같아요...!
패키지 파일도 보니깐 `ESModule` 사용하고 있어서 아마 확장자를 제거할거면 패키지 파일에서 모듈을 `commonjs`로 바꾸고 `import`문도 `require`로 바꿔야 하지 않을까싶네요🙃


<img width="1199" alt="스크린샷 2024-09-20 오전 10 24 25" src="https://github.com/user-attachments/assets/ffdc86f0-353e-467f-8fd3-2ad3812a1e25">

jiyeeeah commented:
@wo-o29 
와.. 이것도 몰랐네요 하하 좀 더 찾아봐야겠어요..! 정말 감사드립니다 :relaxed: 덕분에 정말 많이 배웠습니다!

teagu123 commented on src/utils/generateRandomNum.js:
```diff
@@ -0,0 +1,20 @@
+ */
+export async function generateRandomNum(min, max, length) {
+  const COMPUTER_NUMS = [];
+  while (COMPUTER_NUMS.length < length) {
+    const number = await MissionUtils.Random.pickNumberInRange(min, max);
+    // 중복되는 숫자 없도록 예외 처리
+    if (!COMPUTER_NUMS.includes(number)) {
+      COMPUTER_NUMS.push(number);
+    }
+  }
```
저의 개인적인 코드 스타일이지만 저는 조건문 if문안에 include같은 반환하는 함수를 넣는것이 가독성 부분에서 좋아보이지 않아서 let isOverlap = COMPUTER_NUMS.includes(number)을 반환한후 조건문에 isOverlap을 넣는것이 변수명으로 저 조건문이 어떤 동작을 할수있는지 알수있다고 생각해서 이런 방법도 추천드립니다!!~!!

jiyeeeah replied:
오 가독성 좋아지는 팁 중 하나네요! 좋은 것 같아용 감사합니다 :relaxed:

teagu123 commented on src/utils/validation.js:
```diff
@@ -0,0 +1,27 @@
+ * return 값은 없으며, 유효하지 않을 경우 throw 문을 통해 예외를 발생시키고 프로그램을 종료합니다.
+ */
+export function isInputValid(numString, length) {
+  const INPUT_NUMS = numString.split("");
+  if (INPUT_NUMS.length !== length) {
+    throw new Error(`[ERROR] 입력한 값이 ${length} 자리가 아닙니다.`);
+  }
```
확장성을 생각하면서 코드를 작성하신거같아서 좋았습니다!

질문이 있습니다!
split을 사용하고 length를 하신 이유가 궁금합니다. 일반 string length를 찍어도 동일하게 3이 나오지 않을까 생각해서 질문드립니다. 
또한 위와 같이 진행하면 '1 3'이런식으로 중간에 스페이스를 입력했을때도 3자리로 입력이 된다고 생각합니다! 그래서 trim을 통해서 여백을 줄이고 하는 방법도 좋아보입니다!! 저는 그렇게 진행했었습니다~!

wo-o29 replied:
중간 공백은 `trim`으로 제거가 안되는 걸로 알아서 `replace`, `filter`, `isNaN`같은 걸로 체크해봐도 좋을 것 같아요! 
저도 문제 풀 때 중간 공백은 생각했는데 처음과 끝 공백은 생각 못해서 태기님 코드보고 리팩토링 해야겠다고 생각했네요🙃

jiyeeeah replied:
사실 `split`하고 나서 `length`를 사용하지 않고 `string`에 `length`를 찍어도 되겠지만 가독성때문에 저렇게 작성했던 것 같아요! `INPUT_NUMS`를 먼저 두고 그걸로 유효성 채크를 두 번 해주는 flow를 보여주려고 했던 것 같습니다!

`trim`..! 저도 공백은 생각못했어요..! 좀 더 꼼꼼하게 생각해보고 리팩토링 해보겠습니다! 넘 감사드려요! :smile:

teagu123 commented on src/BaseballGame.js:
```diff
@@ -0,0 +1,74 @@
+   */
+  async startGame() {
+    await MissionUtils.Console.print("숫자 야구 게임을 시작합니다.");
+    this.setComputerNums(await generateRandomNum(1, 9, this.#COMPUTER_NUM_LEN));
```
확장성을 생각하신거같아서 좋습니다! 😮😮

teagu123 commented on src/BaseballGame.js:
```diff
@@ -0,0 +1,74 @@
+
+      this.setUserNums(USER_INPUT);
+
+      const { strike, ball } = await getResult({
+        computerNums: this.#ComputerNums,
+        userNums: this.#UserNums,
+      });
+      await printResult({ strike, ball });
```
print하는 함수와 strike or ball을 카운트 하는 함수를 나누신 이유가 궁금합니다!!
또한 개인적으로 이해했을때는 getResult하는 utils에 있는 파일은 결과를 구한다 라는 의미가 강해서 getRusult파일에서 getResult 함수 마지막에 printResult 함수를 호출하는 어떠신가요?? 로직이 연결되는거다보니 그렇게 진행하는것도 좋을거같아서 말씀드려봅니다 ㅎㅎ

jiyeeeah replied:
카운트 함수와 print하는 함수의 성격이 다르다고 생각해서 나눴습니다! 카운트하는 함수와 print하는 함수의 성격이나 쓰임이 비슷하긴 한 것 같아서  `getResult` 파일 하나에 두 함수를 담아두긴 했습니다!

teagu123 commented on src/BaseballGame.js:
```diff
@@ -0,0 +1,74 @@
+  /**
+   * 게임의 전체 진행을 위한 메서드입니다.
+   */
+  async runGame() {
+    await this.startGame();
+    await this.playGame();
+    await this.endGame();
+  }
```
클래스를 잘쓰시네요 👍👍👍
다른 사람이 코드를 봐도 알아보기 좋고 유지보수에 좋아보입니다!!

jiyeeeah replied:
감사합니다! :smile:

jiyeeeah committed:
[bugfix] 확장자 수정

- import시 확장자 없어서 오류 발생
- 확장자 추가하여 오류 수정

jiyeeeah committed:
[chore] 불필요한 async await 제거

- 동기 함수에 붙여 놓은 async 또는 await 일괄 제거함

jiyeeeah committed:
[chore] export -> export default

- 함수 한개만 내보내는 파일에서 export default 사용하도록 수정

jiyeeeah committed:
[refactor] isInputValid 리팩토링

- 인자로 받은 numString만으로 length와 isNaN 등을 사용할 수 있기 때문에 굳이 배열화 안해도 됨

jiyeeeah committed:
[refactor] generateRandomNum 인자 수정

- 다른 util 함수 (getResult)와 통일성을 위해 generateRandomNum 함수도 인자를 객체로 받도록 수정함

jiyeeeah committed:
[refactor] 변수명 변경

- 상수 역할이 아닌데 스네이크 케이스로 되어있던 변수명 카멜 케이스로 변경
- 기타 변수 또는 아규먼트 명이 특성을 더 잘 드러내도록 수정함 (ex. numString -> input)

jiyeeeah committed:
[refactor] getResult 리팩토링

- getResult 내의 이중 반복문 수정
- continue 활용하여 스트라이크인 경우 넘김
- includes로 볼 개수 카운트

jiyeeeah committed:
[refactor] map 메서드 활용

- map메서드에 콜백함수 말고 Number만 넣어줘도 안의 요소들 형변환 해줌

jiyeeeah committed:
[refactor] includes() 메서드 상수화

- includes() 메서드를 통해 반환되는 값을 상수에 담아두고 조건문에 사용함
